### ISSUE TITLE
`ViewBox` should support "center of zoom" configuration

### ISSUE THREAD

**Issue by goodboy** [pyqtgraph/pyqtgraph#1258]
Currently [`ViewBox.wheelEvent()` only supports zooming "into" or "around" wherever the mouse coordinates are currently located](https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/graphicsItems/ViewBox/ViewBox.py#L1229)  within the view.

There use cases (such as mine) where one might want to keep an alternative coordinate "fixed" during zoom. There should be an easy way to configure the "center" of the zoom functionality.

You can see my commit here that accomplishes this through and override of the method:
https://github.com/pikers/piker/commit/6a6facb6aeacb0b23b5c3ba02b4299b6b06112e1#diff-ff705407e85fb5332ee724f6c0924b16R280

I propose adding a `ViewBox` method or attribute that be overloaded/provided that is called to determine the current *center* for zoom.

**Comment by goodboy** [comment]
Also if there is already a better way to accomplish this I am eager to be enlightened :smile_cat:

**Comment by goodboy** [comment]
@2xB @j9ac9k how do you both feel about a method to register a *zoom focus* callback?
Say something along the lines of:
```python

def zoom_to_right(viewbox):
    """Return the furthest right coordinate in the view.""" 
    return viewbox.boundingRect().topRight()


vb = ViewBox()
vb.set_zoom_focus_callback(zoom_to_right)
```

**Comment by 2xB** [comment]
I for myself do like the idea and would be happy about a Pull Request, maybe even with the possibility to set two callbacks for zoom via dragging right-click and for zoom via mouse wheel.

Just a minor thing: For consistency with the pyqtgraph codebase and Qt, it'd be good to use CamelCase as in `setZoomFocusCallback`.

**Comment by goodboy** [comment]
@2xB sounds good and I'd be happy to make the PR :smile_cat: 

@j9ac9k any thoughts?

**Comment by 2xB** [comment]
Great! In that case, I'd also note here that it'd be great to have one of the given examples extended to show this functionality (maybe customPlot.py?).

**Comment by ixjlyons** [comment]
Is there a use case for dynamically changing the zoom center in a callback versus just setting it e.g. `setZoomCenter(pos)`?

**Comment by goodboy** [comment]
@ixjlyons afaiu the current implementation effectively requires a callback since it's using the cursor's [current position](https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/graphicsItems/ViewBox/ViewBox.py#L1229) (i.e. the `ev.pos()` call).

My use case is similar but I [look up the current "furthest right" x coord on the view](https://github.com/pikers/piker/commit/6a6facb6aeacb0b23b5c3ba02b4299b6b06112e1#diff-ff705407e85fb5332ee724f6c0924b16R280) which I guess in theory you could keep track of in the equivalent "panning callback" and then always look up the last computed value; I'm not sure if that's more sane or not. I guess you would do the same thing with the default *scroll around center* approach but you'd now have to always update some attribute with that on cursor moves. Not sure how much better that is then calling a function.

I'm actually totally down for having a map of cursor or view properties that are continuously updated for this purpose though I wonder if that's needless overhead in a lot of cases?

**Comment by goodboy** [comment]
@2xB yeah what I'd really like is an easy way to switch between scrolling *modes* as well, eg. being able to ctl-scroll to enable the default behavior.

**Comment by j9ac9k** [comment]
I would be +1 for the PR, but given how ...sensitive the viewbox can be...would want to do a fair amount of testing...

**Comment by goodboy** [comment]
> would want to do a fair amount of testing...

Of course. I need to get gud with Qt testing anyway.
Seems like I'm the only peep with a need for this so I'll do a bunch of building around what I've got and then I'll work at factoring it upwards.

**Comment by j9ac9k** [comment]
You can create a draft pull request with some minimum changes, and we can evaluate as you go that way.

**Comment by 2xB** [comment]
Just came to think of this: One could incorporate @ixjlyons concept e.g. by a parameter `fixedZoomCenter`, as e.g. in:
```python3
vb.setFixedZoomCenter((None, None)) # both axes' zoom center controlled by mouse
vb.setFixedZoomCenter((0.5, 0.5)) # both axes' zoom center in middle
vb.setFixedZoomCenter({"wheel": (0.5, 0.5)}) # both axes' zoom center in middle for zoom with mouse wheel
vb.setFixedZoomCenter({"wheel": (0.5, None)}) # x axis zoom center in middle for zoom with mouse wheel, y axis zoom center controlled by mouse position
```
Would avoid having to construct a callback function.

**Comment by goodboy** [comment]
@2xB just clarifying, this api will still require the callbacks underneath right?

**Comment by 2xB** [comment]
@goodboy I'm afraid I don't understand the question. This suggestion would still require the existence of functions for event handling e.g. a `vb.wheelEvent` function, if that's what you mean. Other than that, it would not require any additional callback function.

**Comment by goodboy** [comment]
> Is there a use case for dynamically changing the zoom center in a callback versus just setting it e.g. 

@ixjlyons  @2xB my use case requires looking up a point *in view* which may not be static.

For example say I'm updating a curve in realtime and i want the zoom to zoom into the last updated datum. The user can also pan the placement of this last index and so that value needs to be looked up on every zoom event.

**Comment by goodboy** [comment]
Also feel free to assign this to me @j9ac9k if possible.
